---
title: "visium_seurat"
output: html_document
date: "2024-07-03"
---

# load libraries and set data directory
```{r}
library(Seurat)
library(SeuratObject)
library(ggplot2)
library(tidyverse)
```

# make Seurat object
```{r}
# set data directory
data_dir <- "/Users/taylaalbertini/Desktop/R studio/honours_project/placenta_spatial_annotating/outs"

# check data is there
list.files(data_dir) # Should show filtered_feature_bc_matrix.h5

# seu for Seurat
seu <- Load10X_Spatial(data.dir = data_dir, slice='sample1')
View(seu)
View(seu@meta.data)

# View image in Seurat object
SpatialPlot(seu, pt.size.factor = 1.3, alpha = 1) + 
  NoLegend() + 
  ggtitle('my Visium data') + 
  theme(plot.title = element_text(hjust = 0.5))
```

# data QC and violin plot
```{r}
## nCount = no of molecules / transcripts
## nFeature = no of genes
## MT = mitochondrial

# calculate % MT content per spot
# capital MT for human data. Lower case mt for mouse
seu <- PercentageFeatureSet(seu, "^MT-", col.name = "percent_mito")

head(seu) # a new column should now be there called "percent_mito"

# Vln plot QC data
VlnPlot(seu, features = c("nCount_Spatial", "nFeature_Spatial", "percent_mito"), 
        pt.size = 0.1, 
        group.by = 'orig.ident') + NoLegend()

seu # check dimensions of object

## QC Trim objects individually
# filter data. Change these values based on your data and the violin plots
seu_fil <- subset(seu, subset = nFeature_Spatial > 400 & 
                    nFeature_Spatial < 10000 & 
                    percent_mito < 5)

# Vln again after filtering
VlnPlot(seu_fil, features = c("nCount_Spatial", "nFeature_Spatial", "percent_mito"), 
        pt.size = 0.1, 
        group.by = 'orig.ident') + NoLegend()

## see where there are high counts in the image 
SpatialFeaturePlot(seu, features = "nCount_Spatial") +
  SpatialFeaturePlot(seu_fil, features = "nCount_Spatial")

seu_fil
```

## cluster cells from sketch 3200 cells
```{r}
## Seurat Workflow
seu_fil <- NormalizeData(seu_fil) ## normalise the data
seu_fil <- FindVariableFeatures(seu_fil) ## find the variable features
seu_fil <- ScaleData(seu_fil) ## scale the data

## select 3200 cells and create a new "sketch" assay
sketch <- SketchData(seu_fil, ncells = 3200, method = "LeverageScore", sketched.assay = "sketch")
## set the default assay as "sketch"
DefaultAssay(sketch) <- "sketch"

sketch <- FindVariableFeatures(sketch)
sketch <- ScaleData(sketch)
sketch <- RunPCA(sketch)
sketch <- FindNeighbors(sketch, dims = 1:50)
sketch <- FindClusters(sketch, resolution = 0.5)
sketch <- RunUMAP(sketch, return.model = T, dims = 1:50)
DimPlot(sketch, label = T)

View(sketch@meta.data)
## project the cell clusters onto the full cells 
sketch <- ProjectData(sketch, 
                      assay = "Spatial", 
                      full.reduction = "full.pca",
                      sketched.assay = "sketch",
                      sketched.reduction = "pca",
                      umap.model = "umap",
                      dims = 1:50,
                      refdata = list(seurat_clusters.full = "seurat_clusters"))
View(sketch@meta.data)

## visualise the full clusters based on their spatial location
DefaultAssay(sketch) <- "Spatial" ## set spatial as the default
Idents(sketch) <- sketch@meta.data$seurat_clusters.full

DimPlot(sketch, label = T) ## now have the cell numbers for each cell for the full dataset
SpatialDimPlot(sketch, label = T, repel = T, label.size = 4) ## see the cell clusters on the tissue

## use clusters 0, 1 and 3 to make a new dim plot
cells <- CellsByIdentities(sketch, idents = c(0, 1, 2))
SpatialDimPlot(sketch, cells.highlight = cells[setdiff(names(cells), "NA")],
               cols.highlight = c("yellow", "grey50"),
               facet.highlight = T, combine = T)

## can use the spatial feature plot function to look at the spatial expression of a specific gene
SpatialFeaturePlot(sketch, features = "ACTG2")
## visualise myometrium, decidua and placenta cell gene markers in the cluster
FeaturePlot(object = sketch, 
            features = c("DES", "PRL", "CGA"), label = T, raster = FALSE)

## can find and visualise the top gene expression markers for each cluster
## create downsampled object to make visualisation easier
DefaultAssay(sketch) <- "Spatial"
Idents(sketch) <- "seurat_clusters"
object_subset <- subset(sketch, cells = Cells(sketch[["Spatial"]]), downsample = 1000)

# Order clusters by similarity
DefaultAssay(object_subset) <- "Spatial"
Idents(object_subset) <- "seurat_clusters"
object_subset <- BuildClusterTree(object_subset, assay = "Spatial", reduction = "full.pca", reorder = T)
View(sketch@meta.data)

markers <- FindAllMarkers(object_subset, assay = "Spatial", only.pos = TRUE)
markers %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>%
  slice_head(n = 5) %>%
  ungroup() -> top5

object_subset <- ScaleData(object_subset, assay = "Spatial", features = top5$gene)
p <- DoHeatmap(object_subset, assay = "Spatial", features = top5$gene, size = 2.5) + theme(axis.text = element_text(size = 5.5)) + NoLegend()
p
```

# spot clusters
```{r}
seu_fil <- SCTransform(seu_fil, assay = "Spatial", verbose = FALSE)
seu_fil <- RunPCA(seu_fil, assay = "SCT", verbose = FALSE)
seu_fil <- FindNeighbors(seu_fil, reduction = "pca", dims = 1:30)
seu_fil <- FindClusters(seu_fil, verbose = FALSE)
seu_fil <- RunUMAP(seu_fil, reduction = "pca", dims = 1:30)

# plot reduced dims and overlay on image
p1 <- DimPlot(seu_fil, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(seu_fil, label = TRUE, label.size = 3)
p1 + p2
```

# gene plots
```{r}
# plot individual genes
# change gene in feature argument. 
# genes might be listed differently based on reference used and organism.
# check how they look with rownames(seu_fil) and paste them in to the features argument the same.
genes <- rownames(seu_fil)
genes_to_plot <- "ADAT2"
if (length(genes_to_plot) > 0) {
  FeaturePlot(seu_fil, features = genes_to_plot)
} else {
  message("Selected genes not found in seu_fil.")
}

head(genes)

p1 <- SpatialFeaturePlot(seu_fil, features = "RORB", pt.size.factor = 1)
p2 <- SpatialFeaturePlot(seu_fil, features = "NCAM1", alpha = c(0.1, 1))
p1 + p2
```

# calculate DEGs
```{r}
# calculate differentially expressed genes (DEG's) per cluster.
# can also use FindAllMarkers to find DEG's for all spot clusters.
# ident is your cluster number.
deg_markers <- FindMarkers(seu_fil, ident.1 = 5, ident.2 = 6)

str(deg_markers)
# Use SpatialFeaturePlot again on the DEG's to annotate.
SpatialFeaturePlot(object = seu_fil, features = rownames(deg_markers)[10:15], alpha = c(0.1, 1), ncol = 3)
```








