---
title: "metabolic_comp"
output: html_document
date: "2024-08-09"
---

# load libraries and set project directory 
```{r}
# load relevant libraries 
library(plyr)
library(reshape2)
library(dplyr)
library(stringr)
library(readxl)
library(readr)
library(magrittr)
library(ggplot2)
library(edgeR)
library(limma)
library(Glimma)
library(RColorBrewer)
library(ComplexHeatmap)
library(ggrepel)

# set project directory
projectDir <- "/Users/taylaalbertini/Desktop/R studio/honours_project/fetal_comparison/readCounts_analysis/rawData"

# set path for data tables 
import_mean_female_cpms_TMM <- file.path(projectDir, "mean_female_cpms_TMM.csv")
import_paper_depleted <- file.path(projectDir, "paperDepletion.csv")
input_genome_gff3_file <- file.path(projectDir, "gencode.v29.chr_patch_hapl_scaff.annotation.gff3")
```

# import data files
```{r}
# import mean female cpms table
ncpms_female_TMM <- read_delim(file = import_mean_female_cpms_TMM,
                                   col_names = TRUE,
                                   delim = ",") %>%
  as.data.frame() %>%
  # remove the first column "...1" that;s input when the file is imported
  dplyr::select(., -...1)

# import gong depleted
paper_depleted <- read_delim(file = import_paper_depleted,
                             col_names = TRUE,
                             delim = ",") %>% 
  dplyr::select(., -Chr)
# change colnames for continuity 
colnames(paper_depleted) <- c("ensembl", "hgnc_symbol", "description", "nCPM")
```

# process female depletion data 
```{r}
# transform the tables
female_depleted <- t(ncpms_female_TMM) %>%
  data.frame() %>%
  tibble::rownames_to_column("ensembl")
## rename column names
colnames(female_depleted) <- gsub("X1", "female_brain_TMM", colnames(female_depleted))
colnames(female_depleted) <- gsub("X2", "female_heart_TMM", colnames(female_depleted))
colnames(female_depleted) <- gsub("X3", "female_liver_TMM", colnames(female_depleted))
colnames(female_depleted) <- gsub("X4", "female_lung_TMM", colnames(female_depleted))
colnames(female_depleted) <- gsub("X5", "female_muscle_TMM", colnames(female_depleted))
colnames(female_depleted) <- gsub("X6", "female_placenta_TMM", colnames(female_depleted))
## remove sex_tissue row
female_depleted <- female_depleted[-1,]
```

# import gencode for hngc_symbol annotation
```{r}
# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(input_genome_gff3_file)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames,
                        gene_type = all_gencode_v29@elementMetadata$gene_type)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(., .keep_all = TRUE)
# which leaves us with the following dimensions where there is a unique combination of ensembl ID, hgnc symbol and chromosome
dim(gencode_v29_gene_id_symbol_chr_biotype) # 64837 3
# this does however, leave us with duplicate ensembl IDs
dim(gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl), ]) # 64792 3
# after generating a table to see the duplicates I can see that this is an X/Y chromosome issue. All genes are assigned to both chromosomes
duplicate_rows <- gencode_v29_gene_id_symbol_chr_biotype[duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl) | duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl, fromLast = TRUE),]
# remove duplicates (we don't use the chromosome information for anything)
gencode <- gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl), ]
```

# annotate hgnc_symbol to female_depleted
```{r}
# left join genecode and female_depleted to add gene signatures 
female_depleted <- dplyr::left_join(female_depleted, gencode, by = join_by(ensembl)) 
female_depleted %<>% dplyr::select(., -c(gene_type, seqnames))
```

# left join female_placenta to paper_depleted by hngc_symbol
```{r}
# left join
merge <- dplyr::left_join(paper_depleted, female_depleted, by = "hgnc_symbol") %>% 
  dplyr::select(., -c(ensembl.x, ensembl.y))

# select for mitochondrial function and polymaine metabolic pathway genes

```

