The following Git repo contains all code needed to reproduce the gene depletion comparative analyses, outputs and visualisations from the honours thesis _Sex-specific transcript depletion in the placenta: A comparative analysis across gestation and fetal tissues_. 

**Abstract**

The placenta, the first organ formed during mammalian development, plays a critical role in the feto-maternal interface and pregnancy success. Proper placental function is essential for normal fetal development, with dysfunction linked to pregnancy complications affecting both immediate and long-term health of mother and her offspring. However, the mechanisms of placental dysfunction are not fully understood. 

Tissue-specific gene expression analysis can reveal unique characteristics essential for the placenta's structure and function, though most studies focus on identifying transcripts that are abundant in the placenta compared to other tissues. In contrast, we aimed to investigate the presence of depleted transcripts in the placenta when compared to other fetal tissues, with a focus on sex-specific differences. In addition, we aimed to compare placental gene expression across different gestational stages to explore the variation in depleted transcripts throughout pregnancy. We obtained a total of 356 RNA-sequencing datasets, 126 of which being fetal brain, heart, liver lung and muscle were obtained from the Sequence Read Archive. A further 75 mid-, 33, early- and 56 term-gestation samples were previously sequences within our laboratory. 

The raw reads were aligned to a sex-informed reference genome and gene-level counts were obtained. We detected placental-depleted transcripts by comparing gene expression across tissues. After selecting for protein coding genes, we applied two normalisation techniques in parallel: (1) the trimmed mean of M-values (TMM) and (2) the median ratio (DESeq2) method. We then ranked tissues by their normalised counts per million (nCPM) and applied the following rigorous selection criteria for genes where placenta of interest (e.g. placenta at mid-gestation) was ranked last; (1) nCPM for the tissue ranked second to last must be >1, (2) among those genes, nCPM2(rank = 2nd last)/nCPM(rank = last) must be > 3 and (3) all criteria must be satisfied in both normalisation techniques applied. The identified depleted transcripts were further analysed for enrichment in biological pathways using gene ontology (GO) analysis.

When compared to other fetal tissues, we identified 208 female depleted and 135 male depleted transcripts in the placenta. GO enrichment analysis revealed enriched terms from the female depleted gene set related to fatty acid oxidation, a significant metabolic process for placental function. Additionally, when we compared gene expression in the placenta across gestation (i.e., early- versus mid-gestation) we identified hundreds of depleted transcripts, some of which were involved in metabolic pathways, including the regulation of carbohydrate metabolism. Our results highlight the unique transcriptome of the placenta, providing valuable insights into the molecular basis of placental function and the associated sexual dimorphisms observed in gene expression. The identified placental transcript depletions vary across gestation, reflecting the differing demands and functions of the placenta throughout pregnancy. Overall, our study emphasises the placentaâ€™s distinct metabolic characteristics and its dynamic metabolic landscape throughout gestation.
