---
title: "norm_combined_brainF"
output: html_document
date: "2024-07-29"
---

# load libraries and set project directory 
```{r}
# load relevant libraries 
library(plyr)
library(reshape2)
library(dplyr)
library(stringr)
library(readxl)
library(readr)
library(magrittr)
library(ggplot2)
library(edgeR)
library(limma)
library(Glimma)
library(RColorBrewer)

# set project directory
projectDir <- "/Users/taylaalbertini/Desktop/R studio/honours_project/fetal_comparison/tissue_depletion/rawData"

# set path for data tables 
import_TMM_depletedF <- file.path(projectDir, "TMM_brainF_depleted.csv")
import_DESq_depletedF <- file.path(projectDir, "DESq_brainF_depleted.csv")
input_genome_gff3_file <- file.path("/Users/taylaalbertini/Desktop/R studio/honours_project/oxygen_threshold/rawData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")
```

# import data table 
```{r}
# TMM_depletedF
TMM_depletedF <- read_delim(file = import_TMM_depletedF,
                            col_names = TRUE,
                            delim = ",") %>%
  as.data.frame() %>%
  # remove the first column "...1" that;s input when the file is imported
  dplyr::select(., -...1)

# DESq_depletedF
DESq_depletedF <- read_delim(file = import_DESq_depletedF,
                             col_names = TRUE,
                             delim = ",") %>%
  as.data.frame() %>%
  # remove the first column "...1" that;s input when the file is imported
  dplyr::select(., -...1)
```

# import gencode
```{r}
# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(input_genome_gff3_file)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames,
                        gene_type = all_gencode_v29@elementMetadata$gene_type)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(., .keep_all = TRUE)
# which leaves us with the following dimensions where there is a unique combination of ensembl ID, hgnc symbol and chromosome
dim(gencode_v29_gene_id_symbol_chr_biotype) # 64837 3
# this does however, leave us with duplicate ensembl IDs
dim(gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl), ]) # 64792 3
# after generating a table to see the duplicates I can see that this is an X/Y chromosome issue. All genes are assigned to both chromosomes
duplicate_rows <- gencode_v29_gene_id_symbol_chr_biotype[duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl) | duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl, fromLast = TRUE),]
# remove duplicates (we don't use the chromosome information for anything)
gencode <- gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl), ]
```

# fulfilment of criteria (1) and (2) for TMM and DESq2 normalisation 
```{r}
# tidy the TMM and DESq2 normalised df, filtering to only including the relevant pieces of information and make specific to normalisation technique
## TMM 
## remove rank column
TMM_depletedF <- TMM_depletedF[,-3]
## rename columns to be specific to TMM normalisation
colnames(TMM_depletedF) <- gsub("tissue", "tissue_r5_TMM", colnames(TMM_depletedF))
colnames(TMM_depletedF) <- gsub("nCPM_5", "nCPM_5_TMM", colnames(TMM_depletedF))
colnames(TMM_depletedF) <- gsub("nCPM_6", "nCPM_brain_TMM", colnames(TMM_depletedF))
colnames(TMM_depletedF) <- gsub("nCPM_ratio", "nCPM_ratio_TMM", colnames(TMM_depletedF))
  
## DESq
## remove rank column
DESq_depletedF <- DESq_depletedF[,-3]
## rename columns to be specific to DESq normalisation
colnames(DESq_depletedF) <- gsub("tissue", "tissue_r5_DESq", colnames(DESq_depletedF))
colnames(DESq_depletedF) <- gsub("nCPM_5", "nCPM_5_DESq", colnames(DESq_depletedF))
colnames(DESq_depletedF) <- gsub("nCPM_6", "nCPM_brain_DESq", colnames(DESq_depletedF))
colnames(DESq_depletedF) <- gsub("nCPM_ratio", "nCPM_ratio_DESq", colnames(DESq_depletedF))
colnames(DESq_depletedF) <- gsub("geneid", "ensembl", colnames(DESq_depletedF))
      
# test to ensure there are no duplicated ensembl names
sum(duplicated(TMM_depletedF$ensembl))
        
# left join TMM_depletedF to DESq_depletedF
brainF_depleted <- dplyr::inner_join(TMM_depletedF, DESq_depletedF, by = join_by(ensembl))
## export for plots
write.csv(brainF_depleted, file = file.path(projectDir, "brainF_depleted.csv"))

# left join genecode and brainF_depleted to add gene signatures 
brainF_depleted_annotated <- dplyr::left_join(brainF_depleted, gencode, by = join_by(ensembl))

# tidy brainF_depleted_annotated
## remove seqnames and genetype columns
brainF_depleted_annotated <- dplyr::select(brainF_depleted_annotated, -seqnames, -gene_type)
## remove the period and numbers thereafter
brainF_depleted_annotated_2 <- tidyr::separate(brainF_depleted_annotated, col = ensembl, into = c("ensembl", "right")) %>%
  dplyr::select(., -right)
```

# create a table of depleted transcripts and their descriptions
```{r}
# install and load libraries for 'ensembldb' and 'AnnotationHub'
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

deps <- c("ensembldb", "AnnotationHub")
sapply(deps, function(x){
  if (!require(x, quietly = TRUE, character.only = TRUE))
    BiocManager::install(x)
})

library(ensembldb)
library(AnnotationHub)
# lets take a snapshot of the AnnotationHub data base
ah <- AnnotationHub(ask = FALSE)
# take a look at the information inside
mcols(ah)
# use the `query()` function to extract the information for humans
ensdb_homo <- query(ah, c("EnsDb", "Homo sapiens"))
# keep only the information for AH95744 (which is the same as your gencode V29)
ensdb_104 <- ensdb_homo[["AH95744"]]
# Extract gene-level information and take a look
genes(ensdb_104, return.type = "data.frame") %>% View()
# now put that gene-level information into a data frame
genes_info <- genes(ensdb_104, return.type = "data.frame")

# left join genes_info with brainF_depleted_annotated_2
## change geneid column name to ensembl
colnames(genes_info) <- gsub("gene_id", "ensembl", colnames(genes_info)) 
## left join the dfs
brainF_depleted_annotated_2 <- dplyr::left_join(brainF_depleted_annotated_2, genes_info, by = join_by(ensembl))
## tidy the df removing all information added excluding gene descriptions 
brainF_depleted_annotated_2 <- dplyr::select(brainF_depleted_annotated_2, -11:-17)
brainF_depleted_annotated_2 <- dplyr::select(brainF_depleted_annotated_2, -12:-15)
```

# order depleted genes based on nCPM_ratio_TMM
```{r}
brainF_ordered <- brainF_depleted_annotated[order(brainF_depleted_annotated$nCPM_ratio_TMM, decreasing = TRUE),]
```

# export brainF_depleted_annotated_2
```{r}
write.csv(brainF_depleted_annotated_2, file = file.path(projectDir, "brainF_depleted_annotated.csv"))
```
